onchange_helper/models/models.py

```diff
@@ -1,51 +1,219 @@
+                        continue
+                    vals[name] = self._convert_to_onchange(record, f, v)
+                if not record.id:
+                    result.append((0, 0, vals))
+                elif vals:
+                    result.append((1, record.id, vals))
+                else:
+                    result.append((4, record.id))
+            return result
+        else:
+            return field.convert_to_onchange(value, record, [field.name])
+
+    def play_onchanges(self, values, onchange_fields=None):
+        """
+        Play the onchange methods defined on the current record and return the
+        changed values.
+        The record is not modified by the onchange.
+
+        The intend of this method is to provide a way to get on the server side
+        the values returned by the onchange methods when called by the UI.
+        This method is useful in B2B contexts where records are created or
+        modified from server side.
+
+        The returned values are those changed by the execution of onchange
+        methods registered for the onchange_fields according to the provided
+        values. As consequence, the result will not contain the modifications
+        that could occurs by the execution of compute methods registered for
+        the same onchange_fields.
+
+        It's on purpose that we avoid to trigger the compute methods for the
+        onchange_fields. These compute methods will be triggered when calling
+        the create or write method. In this way we avoid to compute useless
+        information.
+
+
+        :param values: dict of input value that
+        :param onchange_fields: fields for which onchange methods will be
+        played. If not provided, the list of field is based on the values keys.
+        Order in onchange_fields is very important as onchanges methods will
+        be played in that order.
+        :return: changed values
+
+        This method reimplement the onchange method to be able to work on the
+        current recordset if provided.
+        """
+        env = self.env
         if self:
             self.ensure_one()
-            record_values = self._convert_to_write(self.read()[0])
-        else:
-            record_values = {}
-        for field in self._fields:
-            if field not in all_values:
-                all_values[field] = record_values.get(field, False)
-
-        new_values = {}
-        for field in onchange_fields:
-            onchange_values = self.onchange(all_values, field, onchange_specs)
-            new_values.update(self._get_new_values(values, onchange_values))
-            all_values.update(new_values)
-
-        return {
-            f: v for f, v in all_values.iteritems()
-            if not (self._fields[f].compute and not self._fields[f].inverse)
-            and (f in values or f in new_values)}
+
+        if not isinstance(onchange_fields, list):
```

Comment:
https://github.com/acsone/server-tools/blob/e466d297fa4a524242a7559164072e95e65ce2f0/onchange_helper/models/models.py#L135-L139
From line 135 to 139 the conditions should be inverted. If onchange_fields is None, it will be transformed into [None] and therefore will not be evaluated as False
```suggestion
        if not onchange_fields:
            onchange_fields = values.keys()

        elif not isinstance(onchange_fields, list):
            onchange_fields = [onchange_fields]
```

Reply:
good catch!